sympy/matrices/common.py

```diff
@@ -753,6 +755,40 @@ def size(m):
                         diag_entries[(i + row_pos, j + col_pos)] = m[i, j]
                 row_pos += m.rows
                 col_pos += m.cols
+            elif(isinstance(m,dict)):
+                # in this case we're a dict
+                for key,value in m.items():
+                    r_p = 0
+                    c_p = 0
+                    r_p = 0 if key>0 else -(key)
+                    c_p = key if key>0 else 0
+                    func_arg = 0
+                    while (r_p<diag_rows and c_p<diag_cols):
+                        from inspect import isfunction, getargspec
+
+                        if(isinstance(value,list)):
+                            if(len(value)!=min(diag_rows-r_p,diag_cols-c_p)):
+                                raise ValueError("The size of list provided should match the diagonal size")
+                            for i in range(len(value)):
+                                diag_entries[(r_p,c_p)] = value[i]
+                                r_p += 1
+                                c_p += 1
+                        elif(isfunction(value)):
+                            num_args = len(getargspec(value).args)
```

Comment:
import _getnargs from sympy.core.function, then: `num_args = _getnargs(value)`

Reply:
I'm having some problem in using it that way. It returned None. Can you please help me out.

When I use python version 2.7
```
Python 2.7.15+ (default, Feb  3 2019, 13:13:16) 
[GCC 8.2.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy.core.function import _getnargs
>>> def func(a,b,c):
...     return a+b+c
... 
>>> a = _getnargs(func)
>>> print(a)
None
>>> a = _getnargs(lambda i, j: i + 1)
>>> print(a)
None
>>> 
>>> 
>>> 
>>> from inspect import isfunction, getargspec
>>> a = num_args = len(getargspec(func).args)
>>> print(a)
3
>>> a = num_args = len(getargspec(lambda i, j: i + 1).args)
>>> print(a)
2
>>> 
```
Python version 3.7
```
Python 3.7.2+ (default, Feb  2 2019, 14:31:48) 
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy.core.function import _getnargs
>>> def func(a,b,c):
...     return a+b+c
... 
>>> a = _getnargs(func)
>>> print(a)
None
>>> a = _getnargs(lambda i, j: i + 1)
>>> print(a)
None
>>> 
>>> 
>>> 
>>> from inspect import signature
>>> a = len(signature(func).parameters)
>>> print(a)
3
>>> a = len(signature(lambda i, j: i + 1).parameters)
>>> print(a)
2
>>> 
>>> 

```

Reply:
> Can you please help me out.

```diff
diff --git a/sympy/core/function.py b/sympy/core/function.py
index 74fd41c..23aa952 100644
--- a/sympy/core/function.py
+++ b/sympy/core/function.py
@@ -109,6 +109,11 @@ def _getnargs(cls):
             return _getnargs_old(cls.eval)
         else:
             return _getnargs_new(cls.eval)
+    elif inspect.isfunction(cls):
+        if sys.version_info < (3, ):
+            return len(inspect.getargspec(cls).args)
+        else:
+            return len(inspect.signature(cls).parameters)
     else:
         return None

diff --git a/sympy/core/tests/test_function.py b/sympy/core/tests/test_function.py
index 130e19b..7937643 100644
--- a/sympy/core/tests/test_function.py
+++ b/sympy/core/tests/test_function.py
@@ -5,7 +5,7 @@
         Matrix, Basic)
 from sympy.utilities.pytest import XFAIL, raises
 from sympy.core.basic import _aresame
-from sympy.core.function import PoleError, _mexpand
+from sympy.core.function import PoleError, _mexpand, _getnargs
 from sympy.core.sympify import sympify
 from sympy.sets.sets import FiniteSet
 from sympy.solvers.solveset import solveset
@@ -18,6 +18,7 @@
 f, g, h = symbols('f g h', cls=Function)
 _xi_1, _xi_2, _xi_3 = [Dummy() for i in range(3)]

+
 def test_f_expand_complex():
     x = Symbol('x', real=True)

@@ -158,6 +159,15 @@ def test_nargs():
     raises(ValueError, lambda: Function('f', nargs=()))


+def test__getnargs():
+    f = lambda x, y: 1
+    assert _getnargs(f) == 2
+    def f(x, y):
+        pass
+    assert _getnargs(f) == 2
+    assert _getnargs(log) == (1, 2)
+
+
 def test_Lambda():
     e = Lambda(x, x**2)
     assert e(4) == 16
```

Reply:
Thank you, I've made the changes.
Can you please review it now?